### PR TITLE
Calculate base, fee, and quote amounts using totals instead of events

### PR DESCRIPTION
**Pull request by mwoods-figure**
Remove the explicit collection of (potentially unbounded) event data to compute running remaining base, quote and fee amounts, and instead maintain explicit `Uint128` counters for the values. This has the effect of making bid orders take up a deterministic amount of storage space.

Update: tests updated and passing

See https://app.shortcut.com/figure/story/218399/smart-contract-improvement-ideas

@brashidnia-figure Regarding migration of existing bid data, how is that managed?

### CODE REVIEW

**Comment by leeduan** [comment]
> Remove the explicit collection of (potentially unbounded) event data to compute running remaining base, quote and fee amounts, and instead maintain explicit `Uint128` counters for the values. This has the effect of making bid orders take up a deterministic amount of storage space.
> 
> Update: tests updated and passing
> 
> See https://app.shortcut.com/figure/story/218399/smart-contract-improvement-ideas
> 
> @brashidnia-figure Regarding migration of existing bid data, how is that managed?

need to probably add migration code to take the individual events to generate the new fields on existing order book instances during migrate entrypoint?

File: src/contract.rs [hunk 1]
```diff
@@ -178,7 +178,9 @@ pub fn execute(
                     amount: size,
                     denom: base,
                 },
-                events: vec![],
+                remaining_base: Uint128::zero(),
```
Comment (leeduan) [inline src/contract.rs:181]: the way this variable is named, it reads like this should start at base size, quote size, fee size and then reduce per event but i think the way it's structured, it's the opposite of that?

File: src/error.rs [hunk 2]
```diff
@@ -63,6 +63,9 @@ pub enum ContractError {
     #[error("Total (price * size) must be an integer")]
     NonIntegerTotal,
 
+    #[error("Integer overflow")]
+    OverflowError(#[from] cosmwasm_std::OverflowError),
```
Comment (leeduan) [inline src/error.rs:67]: when does this happen? is this related to the `checked_add`? doesnt that return None of overflows?

File: src/contract.rs (hunk 1 above)
Comment (mwoods-figure) [reply src/contract.rs:181]: I was thinking about that after I finished the PR. It would be more appropriate to call those `accumulated_{base,fee,quote}`.

File: src/error.rs (hunk 2 above)
Comment (mwoods-figure) [reply src/error.rs:67]: This can happen with any integer type in Rust. In debug builds overflow will trigger a panic, unless the add is checked. Release won't panic, but will do something (overflow, something else, etc). Since this is using cosmwasm types and not native Rust types, I'm not sure what the behavior will be, but if there's overflow (unlikely, since it's 128 bits), it probably should be treated as an error. The alternative is to use a saturating add.

File: src/bid_order.rs [hunk 3]
```diff
@@ -115,27 +108,7 @@ impl BidOrderV2 {
...
-                                Some(fee) => fee.amount,
-                            },
-                            Action::Reject { fee, .. } => match fee {
-                                None => Uint128::zero(),
-                                Some(fee) => fee.amount,
-                            },
-                        })
-                        .sum::<Uint128>()
-            }
+            Some(fee) => fee.amount - self.remaining_fee,
```
Comment (brashidnia-figure) [inline src/bid_order.rs:111]: would this just be `self.remaining_fee`?

File: src/bid_order.rs [hunk 4]
```diff
@@ -91,16 +93,7 @@ impl BidOrderV2 {
...
-            - self
-                .events
-                .iter()
-                .map(|event| match &event.action {
-                    Action::Fill { base, .. } => base.amount,
-                    Action::Reject { base, .. } => base.amount,
-                    _ => Uint128::zero(),
-                })
-                .sum::<Uint128>()
+        self.base.amount - self.remaining_base
```
Comment (brashidnia-figure) [inline src/bid_order.rs:96]: would this just be `self.remaining_base`?

**Comment by brashidnia-figure** [comment]
> > Remove the explicit collection of (potentially unbounded) event data to compute running remaining base, quote and fee amounts, and instead maintain explicit `Uint128` counters for the values. This has the effect of making bid orders take up a deterministic amount of storage space.
> > Update: tests updated and passing
> > See https://app.shortcut.com/figure/story/218399/smart-contract-improvement-ideas
> > @brashidnia-figure Regarding migration of existing bid data, how is that managed?
> 
> need to probably add migration code to take the individual events to generate the new fields on existing order book instances during migrate entrypoint?

There is logic inside that works around specific versions, but i feel like we could remove all the old ones and only support migrating from v0.17.x or greater, and remove all the old stuff.

To my knowledge, there aren't like any old order books, and if they were, they could:
1.  upgrade from the old version -> v0.18.0 (since all the legacy migration code exists in there)
2.  upgrade from v0.18.0 could upgrade to this new version (only supports latest version upgrade)

Not sure if there is any reason against this, but it could also cut down on the contract size

**Comment by brashidnia-figure** [comment]
> > > > Remove the explicit collection of (potentially unbounded) event data to compute running remaining base, quote and fee amounts, and instead maintain explicit `Uint128` counters for the values. This has the effect of making bid orders take up a deterministic amount of storage space.
> > > > Update: tests updated and passing
> > > > See https://app.shortcut.com/figure/story/218399/smart-contract-improvement-ideas
> > > > @brashidnia-figure Regarding migration of existing bid data, how is that managed?
> > > 
> > > 
> > > need to probably add migration code to take the individual events to generate the new fields on existing order book instances during migrate entrypoint?
> > 
> > 
> > There is logic inside that works around specific versions, but i feel like we could remove all the old ones and only support migrating from v0.17.x (or whatever first uses bidorderv2) or greater, and remove all the old stuff.
> > To my knowledge, there aren't like any active old order books (maybe dlob, which will be replaced with the Fiat Ramp version soon), and if they were, they could:
> > 
> > 1. upgrade from the old version -> v0.18.0 (since all the legacy migration code exists in there)
> > 2. upgrade from v0.18.0 could upgrade to this new version (only supports latest version upgrade)
> > 
> > Not sure if there is any reason against this, but it could also cut down on the contract size
> 
> +1 on maybe time to remove old migrations - does adding non-null fields to a struct that has existing data deserialize correctly with a default value or does this also require a new version of the order struct?

I have not tested it out tbh, but i think adding a non-null field wouldn't work because it would be `None / null` in the actual contract storage

File: src/bid_order.rs (hunk 4 above)
Comment (mwoods-figure) [reply src/bid_order.rs]: it's `self.base.amount - self.accumulated_base` now

File: src/bid_order.rs (hunk 3 above)
Comment (mwoods-figure) [reply src/bid_order.rs]: it's `self.base.amount - self.accumulated_fee` now

**Review by scirner22 — COMMENTED** [review]
Code looks good. I'll defer to Bobak and Lee for migration strategy.

File: src/bid_order.rs [hunk 5]
```diff
@@ -44,7 +44,9 @@ pub struct BidOrderV1 {
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
 pub struct BidOrderV2 {
```
Comment (brashidnia-figure) [inline src/bid_order.rs:45]: I didnt see this earlier, but if you want to be able to perform a migration of the old bid orders (if any exist), you need to create a new struct, `BidOrderV3` with the `accumulated_X` fields

then in the `migrate()` entry point, you need to do this flow:
1. load in the `BidOrderV2` formatted orders from the contract storage (can delete)
2. convert each `BidOrderV2` into a `BidOrderV3` struct
3. write each new `BidOrderV3` order to contract storage

To allow for this backwards compatibility, you need to keep the previous version of the struct, `BidOrderV2` in the contract, and then add the new version `BidOrderV3`

this is basically why there are so many old versions. Each one allows for migrations

Comment (mwoods-figure) [reply src/bid_order.rs:45]: @brashidnia-figure Yes, I started on this path after we talked, which prompted me to open #111 (I figured it was better to clean up the old structs first before adding a new one)

File: src/ask_order.rs [hunk 6]
```diff
@@ -73,125 +53,71 @@ pub fn migrate_ask_orders(
...
+    use super::migrate_ask_orders;
     use crate::error::ContractError;
     use crate::msg::MigrateMsg;
+    use crate::version_info::{set_version_info, VersionInfoV1, CRATE_NAME};
+    use provwasm_mocks::mock_dependencies;
 
     #[test]
-    #[allow(deprecated)]
-    pub fn migrate_legacy_ask_to_v1() -> Result<(), ContractError> {
+    pub fn ask_migration_version_check() -> Result<(), ContractError> {
```
Comment (brashidnia-figure) [inline src/ask_order.rs:72]: nit: separate unit test for each thing being tested

File: src/bid_order.rs [hunk 7]
```diff
@@ -149,714 +170,235 @@ impl BidOrderV2 {
...
     Ok(response)
 }
 
-#[allow(deprecated)]
-pub fn get_legacy_bid_storage(storage: &mut dyn Storage) -> Bucket<BidOrder> {
-    bucket(storage, NAMESPACE_ORDER_BID)
-}
-
-pub fn get_bid_storage(storage: &mut dyn Storage) -> Bucket<BidOrderV2> {
+pub fn get_bid_storage<T>(storage: &mut dyn Storage) -> Bucket<T>
```
Comment (brashidnia-figure) [inline src/bid_order.rs:253]: fancy

File: src/bid_order.rs [hunk 8]
```diff
@@ -907,24 +449,192 @@ mod tests {
...
+                denom: "quote_2".to_string(),
+            },
+        };
+
+        // Store:
+        bid_order_v2_storage.save(&bid1.id.as_bytes(), &bid1)?;
+        bid_order_v2_storage.save(&bid2.id.as_bytes(), &bid2)?;
+
+        // Migrate:
+        migrate_bid_orders(
```
Comment (brashidnia-figure) [inline src/bid_order.rs:576]: check the result/response object?

File: src/bid_order.rs [hunk 9]
```diff
@@ -149,714 +170,235 @@ impl BidOrderV2 {
...
-                bid_fee_rate: None,
-                bid_fee_account: None,
-                ask_required_attributes: None,
-                bid_required_attributes: None,
+        // Contract minimum version:
+        set_version_info(
+            &mut deps.storage,
+            &VersionInfoV1 {
+                definition: CRATE_NAME.to_string(),
+                version: "0.16.2".to_string(), // version too old
```
Comment (brashidnia-figure) [inline src/bid_order.rs:355]: comment is wrong?

Comment (brashidnia-figure) [reply src/bid_order.rs:355]: could also maybe add a greater than 0.16.2 test case

File: src/version_info.rs [hunk 10]
```diff
@@ -92,64 +71,4 @@ mod tests {
...
-                assert_eq!(version_info.version, "0.14.99");
-            }
-            result => panic!("unexpected error: {:?}", result),
-        }
-
-        Ok(())
-    }
-
-    #[test]
-    pub fn get_version_info_unknown() -> Result<(), ContractError> {
```
Comment (brashidnia-figure) [inline src/version_info.rs:133]: why remove this? we could still check that a `ContractError` was returned i think

**Review by brashidnia-figure — APPROVED** [review]
Last thing would be, do we want to bump the version in `Cargo.toml` in here? or separate PR

**Review by brashidnia-figure — APPROVED** [review]
